### DISCUSSION TITLE
Sympy physics mechanics, functions in lambdify

### DISCUSSION THREAD

**Discussion by Peter230655** [sympy/sympy#21774] [Q&A] [answered]
In setting up the equations of motion using Kane‘s method, I was trying to use the sympy.Heaviside function. I used it when I set up the forces list.
My problem comes up when I try to lambdify the rhs = KM.rhs()
The sympy.Heaviside(..) dos not get ‚translated‘ in the np.heaviside(..)
So, when I start the numerical integration, an error comes up.

At the beginning of my little program I do have
from sympy import Heaviside
from numpy import heaviside

How can I ‚translate‘ sympy special functions into their numpy counter parts, so numerical integration will work?

Any help is highly appreciated!

**Comment by moorepants** [comment, marked as answer]
There are functions that don't always work cleanly with lambdify. One option is to use the ability to pass in a custom function for any sympy function/symbol present. This comes from the lambdify docs (https://docs.sympy.org/latest/modules/utilities/lambdify.html):

```python
def mysin(x):
    print('taking the sin of', x)
    return numpy.sin(x)

f = lambdify(x, sin(x), [{'sin': mysin}, 'numpy'])

f(1)
```

**Reply by moorepants** [reply]
The other issue you could be facing is that the integrator can't integrate through the nonlinearity.

**Reply by moorepants** [reply]
It doesn't look like lambdify support Heaviside (yet).

**Comment by Peter230655** [comment]
Worked perfect, thanks!
NB:
1.
In my small example, the discontinuity of heaviside did not negatively
affect the numerical integration.
2.
Your explanation was quickly clear to me. The sympy documentation is more
difficult for me to understand.

On Sat 24. Jul 2021 at 07:37 Jason K. Moore ***@***.***>
wrote:

> There are functions that don't always work cleanly with lambdify. One
> option is to use the ability to pass in a custom function for any sympy
> function/symbol present. This comes from the lambdify docs (
> https://docs.sympy.org/latest/modules/utilities/lambdify.html):
>
> def mysin(x):
>     print('taking the sin of', x)
>     return numpy.sin(x)
> f = lambdify(x, sin(x), [{'sin': mysin}, 'numpy'])
> f(1)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/discussions/21774#discussioncomment-1044575>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT5MQUTRT2DN2GGVYEVBRWTTZJGSBANCNFSM5A5D4QTA>
> .
>
-- 
Best regards,

Peter Stahlecker

**Comment by Peter230655** [comment]
Just for information:
What I am trying to set up is a ‚fully elastic collision‘ between two
bodies.

I used x * Heaviside(x), where x is the ‚penetration‘ of one body into the
other one.
I also used Abs(x+ Abs(x)) which should be equivalent, but numerically
‚better‘, because Heaviside is not continuous at x = 0.

However, numerically integrating both versions showed no obvious
difference, at least I could not see one.


On Sun 25. Jul 2021 at 12:22 Peter Stahlecker ***@***.***>
wrote:

> Worked perfect, thanks!
> NB:
> 1.
> In my small example, the discontinuity of heaviside did not negatively
> affect the numerical integration.
> 2.
> Your explanation was quickly clear to me. The sympy documentation is more
> difficult for me to understand.
>
> On Sat 24. Jul 2021 at 07:37 Jason K. Moore ***@***.***>
> wrote:
>
>> There are functions that don't always work cleanly with lambdify. One
>> option is to use the ability to pass in a custom function for any sympy
>> function/symbol present. This comes from the lambdify docs (
>> https://docs.sympy.org/latest/modules/utilities/lambdify.html):
>>
>> def mysin(x):
>>     print('taking the sin of', x)
>>     return numpy.sin(x)
>> f = lambdify(x, sin(x), [{'sin': mysin}, 'numpy'])
>> f(1)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/sympy/sympy/discussions/21774#discussioncomment-1044575>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AT5MQUTRT2DN2GGVYEVBRWTTZJGSBANCNFSM5A5D4QTA>
>> .
>>
> --
> Best regards,
>
> Peter Stahlecker
>
-- 
Best regards,

Peter Stahlecker
